### DISCUSSION TITLE
Are Graphcache updates sync?

### DISCUSSION THREAD

**Discussion by mattfysh** [urql-graphql/urql#1953] [Q&A] [answered]
I'm seeing some race conditions in my code leading to JS errors.
Given the following config:

```js
updates: {
    Mutation: {
      createInstall: (result, args, cache) => {
        const parentBoard = { __typename: 'Board', id: args.input.boardId }
        const installs = cache.resolve(parentBoard, 'installs')
        const items = cache.resolve(installs, 'items')
        if (Array.isArray(items)) {
          items.push(result.createInstall)
          cache.link(installs, 'items', items)
        }
      },
```

And then code in a **Svelte component** that does:

```js
const { data } = await mutation({query: createInstall})({ input })
```

Will the cache be guaranteed to be updated -- and cache updates distributed to all active queries -- *before* the `await mutation` resolves?

Thanks!

**Comment by kitten** [comment, marked as answer]
Yes, all cache updates are synchronous and directly reactive to results, meaning that they're exectued as soon as possible or needed.

**Reply by mattfysh** [reply]
@kitten I am getting a race condition, and I've gone through my code and can't find anything that could be causing the race condition. Could it possibly be in the `@urql/svelte` package?

The problem code is here:

```js
const { data } = await mutation({query: createInstall})({ input })
// (1) do something with query that should have been updated using Cache Updates
```

At (1) sometimes the query operationStore contains the update, and sometimes it doesn't.

The failure happens ~50% of the runs.

If I introduce an artificial delay to my code, then I no longer get the race condition, e.g:

```js
const { data } = await mutation({query: createInstall})({ input })
setTimeout(() => {
  // (1) do something with query that should have been updated using Cache Updates
}, 200)
```

**Reply by mattfysh** [reply]
Ok that wasn't it... even getting the error with the setTimeout, even though it's occurring less now.

Will keep digging, I think it may be something to do with Svelte's `flush` function. There are two components which are invalidated by the manual Cache Update. If I set a breakpoint inside `flush` and inspect the `dirty_components` array:

1. If the order is `[Board, Sidebar]` then everything works ok
2. If the order is `[Sidebar, Board]` then I get the error.

Not too sure what's going on but will keep you posted if I find an answer!

**Reply by mattfysh** [reply]
Noticed it only seems to be happening if the array being linked was previously empty, e.g.

```
const todos = cache.resolve('Query', 'todos');
if (Array.isArray(todos)) {
  todos.push(result.createTodo);
  cache.link('Query', 'todos', todos);
}
```

If todos is NOT an empty array, then everything works as expected

If todos is an empty array, I get the error where
1. `await mutation` resolves but the active query does not contain the manual Cache Updates
2. in fact, the active query is re-executed and a new network request is sent - not sure why, this doesn't happen if the array wasn't empty

**Reply by mattfysh** [reply]
Alright, I switched off the other component that was querying the same data - and now I can reproduce the problem consistently.

**Reply by mattfysh** [reply]
Ok I've figured out what's happening...

1. The mutation response has overlap with an active query in the same Svelte component
2. When the manual Cache Update occurs, it uses `cache.link` to insert a new object into a list on the active query
3. The new object (result of mutation) does not contain all of the data required by the query
4. This explains why a new network request is created - to fill in the missing pieces of data
5. The `await mutation` resolves without waiting for the result of the re-executed query

I still don't quite understand why the manual Cache Update is missing from the active query, even while waiting for the network response?
